rfc/0075-faster-failover-and-configuration-push.md

```diff
@@ -0,0 +1,315 @@
+feature can still be used as a building block to implement a debouncing mechanism. When properly configured, it can help
+reduce the number of requests. Further details on this topic will be covered in the "Library Changes" section.
+
+## Library Changes
+
+### Configuration Push
+
+The previously mentioned `ClustermapChangeNotificationBrief` feature enables the SDK to subscribe all connections for
+configuration updates. These notifications are lightweight and can be deduplicated by the server when the
+`DedupeNotMyVbucketClustermap` option is negotiated.
+
+#### Mixed Clusters
+
+In clusters where there is a mix of nodes with older server versions, meaning that some nodes do not acknowledge
+`ClustermapChangeNotificationBrief`, the respective connection should notify the configuration monitor about its lack of
+support for configuration pushes from the server. As a result, the monitor should utilize the old polling mechanism for
+this particular node instead.
+
+### Enhancements in Handling the `NotMyVbucket` Status
+
+Combination of `DedupeNotMyVbucketClustermap` and `ClustermapChangeNotificationBrief` allows to save traffic by not
+sending configuration, if SDK already seen the same revision, and also sends only pair of `Epoch`/`Revision`. So it is
+up to SDK to initiate configuration update once the non-empty payload returned along with `NotMyVbucket` status code.
+
+Several modifications are required in the SDK:
+1. The retry orchestrator should be able to retry an operation based on configuration updates rather than the timer signal.
+2. The configuration monitor should have the ability to throttle configuration requests due to the following reasons:
+   1. During rebalance, multiple operations may return a `NotMyVbucket` status, triggering a configuration refresh.
+   2. Since `ClustermapChangeNotificationBrief` will cause all connections to subscribe to updates and receive them, it is
+      necessary to account for potential high volumes of updates.
+
+Below is a diagram that illustrates an example of the SDK workflow, where the GET request is waiting for the arrival of
+a new configuration.
+
+```mermaid
+sequenceDiagram
+    autonumber
+    conn_1->>+kv_node_1: get("foo", vb=115)
+    kv_node_1->>-conn_1: NotMyVbucket(epoch=1, rev=11)
+    conn_1-->>+retry_orchestrator: pending(get, "foo", epoch=1, rev=11)
+    retry_orchestrator-->retry_orchestrator: put operation to wating queue
+    conn_1-->>+config_monitor: refresh configuration
+    config_monitor-->config_monitor: wait to throttle config requests
+    config_monitor->>+conn_2: get_config()
+    conn_2->>+kv_node_2: get_config()
+    kv_node_2->>-conn_2: configuration(epoch=1, rev=11)
+    conn_2->>-config_monitor: apply new configuration
+    config_monitor->>retry_orchestrator: purge waiting queue(epoch=1, rev=11)
+    retry_orchestrator->>conn_1: retry get("foo")
+    conn_1->>+kv_node_2: get("foo", vb=115)
+    kv_node_2->>-conn_1: Success()
+```
```

Comment:
Nb to anyone else wanting to see this visually: https://mermaid.live/edit#pako:eNqNk01vgzAMhv9KlBNo7ADSLkjltOt2qbQTEkqDgaglYYnTClX97zOwIrrSqZwc7Od14o8zl6YEnnIH3x60hHclaivaXDP6hEejfbsDO52l0bqIX7PsZX8sNIFFnLIaMMh5ZUzOI3bcbeL4LZzC5yAiXic2ZZ8GP_qvnZd74qAzstnEEbNwJDC8STPksYC2L4yVDTi0Ao1NWQe6VLoOKHHE5szrUvc8ya6KemSmAzopoxkadiJL14yq4uHuWmRWqi5ao9VIW6gsuIZN__2kMlOLUKL_siehcEiIjTWIB_glSJNyO3SrMtMddJGM9S8mb7CsX7Js02rY7JzbQ2E3L_ivP1fq5i2i6w4903B6ohLZgz7YGsaazNUPnm5tll2nbHQuRzN8NMHJMxOcLCd466UE54KQR7wF2wpV0gKdByLn2EALOU_JLIXd5zzXF4obNmnba8lTtB4i7rtS4HXZeFqJg4PLD_iiNoo

Reply:
![image](https://github.com/couchbaselabs/sdk-rfcs/assets/7280/8a495f96-4c4e-4e2e-89da-001362ff3048)